### PR TITLE
fix(backend): Add more error reasons for refresh token query param

### PR DESCRIPTION
**Pull request by anagstef**
## Description

This PR extracts more error reasons to be set on `__clerk_refresh` query param, instead of falling back to `unexpected-refresh-error`.

New error codes are:
- `missing-api-client`
- `missing-session-token`
- `missing-refresh-token`
- `session-token-decode-failed`
-  `fetch-network-error`

## Checklist

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [x] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

**Comment by anagstef** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @anagstef - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 1.3.7-snapshot.v150a390 |
| @clerk/backend | 1.13.3-snapshot.v150a390 |
| @clerk/chrome-extension | 1.3.9-snapshot.v150a390 |
| @clerk/clerk-js | 5.22.4-snapshot.v150a390 |
| @clerk/elements | 0.15.5-snapshot.v150a390 |
| @clerk/clerk-expo | 2.2.15-snapshot.v150a390 |
| @clerk/express | 0.1.4-snapshot.v150a390 |
| @clerk/fastify | 1.0.46-snapshot.v150a390 |
| @clerk/localizations | 3.0.5-snapshot.v150a390 |
| @clerk/nextjs | 5.6.1-snapshot.v150a390 |
| @clerk/clerk-react | 5.9.2-snapshot.v150a390 |
| @clerk/remix | 4.2.30-snapshot.v150a390 |
| @clerk/clerk-sdk-node | 5.0.43-snapshot.v150a390 |
| @clerk/shared | 2.8.2-snapshot.v150a390 |
| @clerk/tanstack-start | 0.4.6-snapshot.v150a390 |
| @clerk/testing | 1.3.4-snapshot.v150a390 |
| @clerk/themes | 2.1.31-snapshot.v150a390 |
| @clerk/types | 4.21.1-snapshot.v150a390 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@1.3.7-snapshot.v150a390 --save-exact
```
`@clerk/backend`
```sh
npm i @clerk/backend@1.13.3-snapshot.v150a390 --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.3.9-snapshot.v150a390 --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@5.22.4-snapshot.v150a390 --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.15.5-snapshot.v150a390 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@2.2.15-snapshot.v150a390 --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@0.1.4-snapshot.v150a390 --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@1.0.46-snapshot.v150a390 --save-exact
```
`@clerk/localizations`
```sh
npm i @clerk/localizations@3.0.5-snapshot.v150a390 --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.6.1-snapshot.v150a390 --save-exact
```
`@clerk/clerk-react`
```sh
npm i @clerk/clerk-react@5.9.2-snapshot.v150a390 --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.2.30-snapshot.v150a390 --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.43-snapshot.v150a390 --save-exact
```
`@clerk/shared`
```sh
npm i @clerk/shared@2.8.2-snapshot.v150a390 --save-exact
```
`@clerk/tanstack-start`
```sh
npm i @clerk/tanstack-start@0.4.6-snapshot.v150a390 --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.3.4-snapshot.v150a390 --save-exact
```
`@clerk/themes`
```sh
npm i @clerk/themes@2.1.31-snapshot.v150a390 --save-exact
```
`@clerk/types`
```sh
npm i @clerk/types@4.21.1-snapshot.v150a390 --save-exact
```

File: packages/backend/src/tokens/request.ts [hunk 1]
```diff
@@ -195,16 +207,36 @@ ${error.getFullMessage()}`,
   }
 
   async function refreshToken(authenticateContext: AuthenticateContext): Promise<string> {
-    // To perform a token refresh, apiClient must be defined.
-    assertApiClient(options.apiClient);
+    try {
+      // To perform a token refresh, apiClient must be defined.
+      assertApiClient(options.apiClient);
+    } catch (err: any) {
+      throw { errors: [{ code: RefreshTokenErrorReason.MissingApiClient, message: err?.message || '' }] };
```
Comment (LekoArts) [inline packages/backend/src/tokens/request.ts:214]: Why the mix between single `error` and `errors` array? I've only seen single errors, not multiple, in your code below 🤔

Comment (anagstef) [reply packages/backend/src/tokens/request.ts:214]: @LekoArts an `errors` array is returned on API errors.

Comment (anagstef) [reply packages/backend/src/tokens/request.ts]: Just refactored this, to be clearer.

**Comment by anagstef** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @anagstef - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 1.3.7-snapshot.v51c436a |
| @clerk/backend | 1.13.3-snapshot.v51c436a |
| @clerk/chrome-extension | 1.3.9-snapshot.v51c436a |
| @clerk/clerk-js | 5.22.4-snapshot.v51c436a |
| @clerk/elements | 0.15.5-snapshot.v51c436a |
| @clerk/clerk-expo | 2.2.15-snapshot.v51c436a |
| @clerk/express | 0.1.4-snapshot.v51c436a |
| @clerk/fastify | 1.0.46-snapshot.v51c436a |
| @clerk/localizations | 3.0.5-snapshot.v51c436a |
| @clerk/nextjs | 5.6.1-snapshot.v51c436a |
| @clerk/clerk-react | 5.9.2-snapshot.v51c436a |
| @clerk/remix | 4.2.30-snapshot.v51c436a |
| @clerk/clerk-sdk-node | 5.0.43-snapshot.v51c436a |
| @clerk/shared | 2.8.2-snapshot.v51c436a |
| @clerk/tanstack-start | 0.4.6-snapshot.v51c436a |
| @clerk/testing | 1.3.4-snapshot.v51c436a |
| @clerk/themes | 2.1.31-snapshot.v51c436a |
| @clerk/types | 4.21.1-snapshot.v51c436a |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@1.3.7-snapshot.v51c436a --save-exact
```
`@clerk/backend`
```sh
npm i @clerk/backend@1.13.3-snapshot.v51c436a --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.3.9-snapshot.v51c436a --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@5.22.4-snapshot.v51c436a --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.15.5-snapshot.v51c436a --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@2.2.15-snapshot.v51c436a --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@0.1.4-snapshot.v51c436a --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@1.0.46-snapshot.v51c436a --save-exact
```
`@clerk/localizations`
```sh
npm i @clerk/localizations@3.0.5-snapshot.v51c436a --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.6.1-snapshot.v51c436a --save-exact
```
`@clerk/clerk-react`
```sh
npm i @clerk/clerk-react@5.9.2-snapshot.v51c436a --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.2.30-snapshot.v51c436a --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.43-snapshot.v51c436a --save-exact
```
`@clerk/shared`
```sh
npm i @clerk/shared@2.8.2-snapshot.v51c436a --save-exact
```
`@clerk/tanstack-start`
```sh
npm i @clerk/tanstack-start@0.4.6-snapshot.v51c436a --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.3.4-snapshot.v51c436a --save-exact
```
`@clerk/themes`
```sh
npm i @clerk/themes@2.1.31-snapshot.v51c436a --save-exact
```
`@clerk/types`
```sh
npm i @clerk/types@4.21.1-snapshot.v51c436a --save-exact
```

**Comment by clerk-cookie** [comment]
Hey @anagstef - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 1.3.7-snapshot.ve9007ed |
| @clerk/backend | 1.13.3-snapshot.ve9007ed |
| @clerk/chrome-extension | 1.3.9-snapshot.ve9007ed |
| @clerk/clerk-js | 5.22.4-snapshot.ve9007ed |
| @clerk/elements | 0.15.5-snapshot.ve9007ed |
| @clerk/clerk-expo | 2.2.15-snapshot.ve9007ed |
| @clerk/express | 0.1.4-snapshot.ve9007ed |
| @clerk/fastify | 1.0.46-snapshot.ve9007ed |
| @clerk/localizations | 3.0.5-snapshot.ve9007ed |
| @clerk/nextjs | 5.6.1-snapshot.ve9007ed |
| @clerk/clerk-react | 5.9.2-snapshot.ve9007ed |
| @clerk/remix | 4.2.30-snapshot.ve9007ed |
| @clerk/clerk-sdk-node | 5.0.43-snapshot.ve9007ed |
| @clerk/shared | 2.8.2-snapshot.ve9007ed |
| @clerk/tanstack-start | 0.4.6-snapshot.ve9007ed |
| @clerk/testing | 1.3.4-snapshot.ve9007ed |
| @clerk/themes | 2.1.31-snapshot.ve9007ed |
| @clerk/types | 4.21.1-snapshot.ve9007ed |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@1.3.7-snapshot.ve9007ed --save-exact
```
`@clerk/backend`
```sh
npm i @clerk/backend@1.13.3-snapshot.ve9007ed --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.3.9-snapshot.ve9007ed --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@5.22.4-snapshot.ve9007ed --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.15.5-snapshot.ve9007ed --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@2.2.15-snapshot.ve9007ed --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@0.1.4-snapshot.ve9007ed --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@1.0.46-snapshot.ve9007ed --save-exact
```
`@clerk/localizations`
```sh
npm i @clerk/localizations@3.0.5-snapshot.ve9007ed --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.6.1-snapshot.ve9007ed --save-exact
```
`@clerk/clerk-react`
```sh
npm i @clerk/clerk-react@5.9.2-snapshot.ve9007ed --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.2.30-snapshot.ve9007ed --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.43-snapshot.ve9007ed --save-exact
```
`@clerk/shared`
```sh
npm i @clerk/shared@2.8.2-snapshot.ve9007ed --save-exact
```
`@clerk/tanstack-start`
```sh
npm i @clerk/tanstack-start@0.4.6-snapshot.ve9007ed --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.3.4-snapshot.ve9007ed --save-exact
```
`@clerk/themes`
```sh
npm i @clerk/themes@2.1.31-snapshot.ve9007ed --save-exact
```
`@clerk/types`
```sh
npm i @clerk/types@4.21.1-snapshot.ve9007ed --save-exact
```

**Review by nikosdouvlis — APPROVED** [review]

**Pull request merged**